lib/common/socket.c

```diff
@@ -1668,39 +1721,100 @@ int h2o_socket_ebpf_init_key(h2o_ebpf_map_key_t *key, void *_sock)
     return h2o_socket_ebpf_init_key_raw(key, sock_type, (void *)&local, (void *)&remote);
 }
 
-h2o_ebpf_map_value_t h2o_socket_ebpf_lookup(h2o_loop_t *loop, int (*init_key)(struct st_h2o_ebpf_map_key_t *key, void *cbdata),
-                                            void *cbdata)
+static int get_return_map_fd(h2o_loop_t *loop)
 {
-    // try open map if not opened
-    open_tracing_map(loop);
+    static __thread int fd = -1;
+    static __thread uint64_t last_attempt = 0;
 
-    // map is not connected, fallback accepting probe
-    if (tracing_map_fd < 0)
-        return (h2o_ebpf_map_value_t){0};
+    // only check every second
+    uint64_t now = h2o_now(loop);
+    if (last_attempt - now < 1000)
+        return fd;
+
+    last_attempt = now;
+
+    // check if map exists at path
+    struct stat s;
+    if (stat(H2O_EBPF_RETURN_MAP_PATH, &s) == -1) {
+        // map path unavailable, cleanup fd if needed and leave
+        if (fd >= 0) {
+            close(fd);
+            fd = -1;
+        }
+        return -1;
+    }
 
-    // lookup map for our key
+    if (fd >= 0)
+        return fd; // map still exists and we have a fd
+
+    // map exists, try bpf_obj_get
+    union bpf_attr attr = {
+        .pathname = (uint64_t)H2O_EBPF_RETURN_MAP_PATH,
+    };
+    fd = syscall(__NR_bpf, BPF_OBJ_GET, &attr, sizeof(attr));
+    if (fd < 0)
+        h2o_perror("BPF_OBJ_GET failed");
+    return fd;
+}
+
+uint64_t h2o_socket_ebpf_lookup_flags(h2o_loop_t *loop, int (*init_key)(h2o_ebpf_map_key_t *key, void *cbdata), void *cbdata)
+{
+    uint64_t flags = 0;
+
+    int tracing_map_fd = get_tracing_map_fd(loop);
     h2o_ebpf_map_key_t key;
-    if (!init_key(&key, cbdata))
-        return (h2o_ebpf_map_value_t){0};
-    return lookup_map(&key);
+    if ((tracing_map_fd >= 0 || H2O_SOCKET_LOOKUP_FLAGS_ENABLED()) && init_key(&key, cbdata)) {
+        if (tracing_map_fd >= 0)
+            ebpf_map_lookup(tracing_map_fd, &key, &flags);
+
+        int return_map_fd;
+        if (H2O_SOCKET_LOOKUP_FLAGS_ENABLED() && (return_map_fd = get_return_map_fd(loop)) >= 0) {
+            pid_t tid = gettid();
+
+            // make sure a possible old flags is not set,
+            // otherwise the subsequent logic will be unreliable.
```

Comment:
```suggestion
            /* Make sure a possible old flags is not set, otherwise the subsequent logic will be unreliable. */
```
Please see my other comment regarding why.